parser-typechecker/src/Unison/Util/Text.hs

```diff
@@ -0,0 +1,133 @@
+{-# Language BangPatterns #-}
+{-# Language GeneralizedNewtypeDeriving #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Unison.Util.Text where
+
+import Data.String (IsString(..))
+import Data.Foldable (toList)
+import Data.List (unfoldr)
+import Prelude hiding (take,drop,replicate)
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import qualified Unison.Util.Bytes as B
+import qualified Unison.Util.Rope as R
+
+-- Text type represented as a `Rope` of chunks
+newtype Text = Text (R.Rope Chunk) deriving (Eq,Ord,Semigroup,Monoid)
+
+data Chunk = Chunk {-# unpack #-} !Int {-# unpack #-} !T.Text
+
+empty :: Text
+empty = Text mempty
+
+one, singleton :: Char -> Text
+one ch = Text (R.one (chunk (T.singleton ch)))  
+singleton = one
+
+threshold :: Int
+threshold = 512
+
+replicate :: Int -> Text -> Text
+replicate n t | size t * n < threshold = Text (R.one (chunk (T.replicate n (toText t)))) 
+replicate 0 _ = mempty 
+replicate 1 t = t
+replicate n t = 
+  replicate (n `div` 2) t <> replicate (n - (n `div` 2)) t
+
+chunkToText :: Chunk -> T.Text
+chunkToText (Chunk _ t) = t
+
+chunk :: T.Text -> Chunk
+chunk t = Chunk (T.length t) t 
+
+take :: Int -> Text -> Text
+take n (Text t) = Text (R.take n t)
+
+drop :: Int -> Text -> Text
+drop n (Text t) = Text (R.drop n t)
+
+uncons :: Text -> Maybe (Char, Text)
+uncons t | size t == 0 = Nothing
+uncons t = (,drop 1 t) <$> at 0 t
+
+unsnoc :: Text -> Maybe (Text, Char)
+unsnoc t | size t == 0 = Nothing
+unsnoc t = (take (size t - 1) t,) <$> at (size t - 1) t
+
+at :: Int -> Text -> Maybe Char
+at n (Text t) = R.index n t
+
+size :: Text -> Int
+size (Text t) = R.size t
+
+reverse :: Text -> Text
+reverse (Text t) = Text (R.reverse t)
+
+fromUtf8 :: B.Bytes -> Either String Text
+fromUtf8 bs = 
+  case T.decodeUtf8' (B.toByteString bs) of
+    Right t -> Right (fromText t)
+    Left e -> Left (show e)
+
+toUtf8 :: Text -> B.Bytes
+toUtf8 (Text t) = B.Bytes (R.map (B.chunkFromByteString . T.encodeUtf8 . chunkToText) t)
+
+fromText :: T.Text -> Text
+fromText s | T.null s = mempty
+fromText s = go (Text (R.one (chunk s)))
```

Comment:
I think it may be more efficient to use `Text.chunksOf` to split the text into threshold-sized chunks, then you can build a balanced tree directly from that list.

Seems like this version has to `take` and `drop` over the same pieces of text many times whereas `chunksOf` could split into the correct size on the first go 😄

Reply:
Ah, good call!